tests/attacks/inference/test_membership_inference.py

```diff
@@ -0,0 +1,363 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+import pytest
+
+import numpy as np
+import torch.nn as nn
+import torch.optim as optim
+
+import keras
+
+from art.attacks.inference import (
+    MembershipInferenceBlackBoxRuleBased,
+    MembershipInferenceBlackBox,
+)
+from art.estimators.classification.pytorch import PyTorchClassifier
+from art.estimators.classification.keras import KerasClassifier
+from art.estimators.estimator import BaseEstimator
+
+from tests.attacks.utils import backend_test_classifier_type_check_fail
+
+
+logger = logging.getLogger(__name__)
+attack_train_ratio = 0.5
+num_classes_iris = 3
+num_classes_mnist = 10
+
+
+@pytest.fixture()
+def fix_get_mnist_subset(get_mnist_dataset):
+    (x_train_mnist, y_train_mnist), (x_test_mnist, y_test_mnist) = get_mnist_dataset
+    n_train = 1000
+    n_test = 200
+    yield x_train_mnist[:n_train], y_train_mnist[:n_train], x_test_mnist[:n_test], y_test_mnist[:n_test]
+
+
+def test_rule_based_image(fix_get_mnist_subset, get_image_classifier_list_for_attack):
+    classifier_list = get_image_classifier_list_for_attack(MembershipInferenceBlackBoxRuleBased)
+    if not classifier_list:
+        logging.warning("Couldn't perform this test because no classifier is defined")
+        return
+
+    x_train, y_train, x_test, y_test = fix_get_mnist_subset
+
+    for classifier in classifier_list:
+        # print(type(classifier).__name__)
+        attack = MembershipInferenceBlackBoxRuleBased(classifier)
+        # infer attacked feature
+        inferred_train = attack.infer(x_train, y_train)
+        inferred_test = attack.infer(x_test, y_test)
+        # check accuracy
+        # print(inferred_train)
+        # print(inferred_test)
+        train_pos = sum(inferred_train) / len(inferred_train)
+        test_pos = sum(inferred_test) / len(inferred_test)
+        assert (train_pos > test_pos or
+                train_pos == pytest.approx(test_pos, abs=0.03) or
+                test_pos == 1)
+
+
+def test_rule_based_tabular(get_iris_dataset, get_tabular_classifier_list):
+    classifier_list = get_tabular_classifier_list(MembershipInferenceBlackBoxRuleBased)
+    if not classifier_list:
+        logging.warning("Couldn't perform this test because no classifier is defined")
+        return
+
+    (x_train, y_train), (x_test, y_test) = get_iris_dataset
+
+    for classifier in classifier_list:
+        # print(type(classifier).__name__)
+        attack = MembershipInferenceBlackBoxRuleBased(classifier)
+        # infer attacked feature
+        inferred_train = attack.infer(x_train, y_train)
```

Comment:
lines from 90 to 97 are pratically identical across all tests. For maintenance purposes, it would be better to use a `_backend_test_something` method was used. An example of this pattern can be found at `tests.attacks.evasion.test_fast_gradient`

Reply:
So are you saying I should put the _backend_test_something method in attacks/utils.py? Even if it's only used in this test file? Or should I add it within test_membership_inference.py?

Reply:
yes I think leaving it in `test_membership_inference.py` is probably a better idea since it's only used here